### PR TITLE
Add residual monitor for checking for correct handling of essential boundary conditions

### PR DESCRIPTION
**Pull request by v-dobrev**
This branch adds a simple `ResidualBCMonitor` that can be used to check if essential b.c. are properly imposed on the initial guess, rhs, operator, and preconditioner.

To use this class simply add  something like
```c++
   ResidualBCMonitor bc_mon(ess_tdof_list);
   iter_solver.SetMonitor(bc_mon);
```
before the call to `iter_solver.Mult`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1633](https://github.com/mfem/mfem/pull/1633) | @v-dobrev | @tzanio | @barker29 + @jandrej | 07/19/20 | 08/04/20 | 08/13/20 | |

### CODE REVIEW

**Comment by stefanhenneking** [comment]
@v-dobrev  Thanks for adding this, that's a good idea! We'd have caught the issue fixed in https://github.com/mfem/mfem/pull/1563/commits/5c1f9b64af4574c3515e1cda226c5d675279e47d earlier.

**Comment by v-dobrev** [comment]
> @v-dobrev Thanks for adding this, that's a good idea! We'd have caught the issue fixed in [5c1f9b6](https://github.com/mfem/mfem/commit/5c1f9b64af4574c3515e1cda226c5d675279e47d) earlier.

@stefanhenneking, yes, this PR was inspired by that issue and I also tested it there.

**Comment by jandrej** [comment]
LGTM

File: linalg/solvers.cpp [hunk 1]
```diff
@@ -2665,6 +2665,42 @@ void BlockILU::Mult(const Vector &b, Vector &x) const
...
+      bc_norm_squared = glob_bc_norm_squared;
+      int rank;
+      MPI_Comm_rank(comm, &rank);
+      print = (rank == 0);
+   }
+#endif
+   if ((it == 0 || final || bc_norm_squared > 0.0) && print)
+   {
+      mfem::out << "      ResidualBCMonitor : b.c. residual norm = "
+                << sqrt(bc_norm_squared) << std::endl;
```
Comment (barker29) [inline linalg/solvers.cpp:2699]:
```suggestion
                << sqrt(bc_norm_squared) << endl;
```

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**